custom_metrics/wpt_bodies.js

```diff
@@ -0,0 +1,1366 @@
+            }
+          }
+          return {characters: text.length, words: words ?? 0};
+        }
+        
+        function processHeadings(d) {
+          let target = {first_non_empty_heading_hidden: false};
+          for(let l=1; l < 9; l++) {
+            let nodes = Array.from(d.querySelectorAll('h'+l));
+
+            let characters = 0;
+            let words = 0;
+            // if don't have a primary heading yet, search for one.
+            
+            var hs = nodes.map(n => {
+              let h = processHeading(d, target, n, l);
+              characters += h.characters;
+              words += h.words;
+              return h;
+            });
+      
+            target["h"+l] = {
+              total: nodes.length,
+              non_empty_total: nodes.filter(e => seoText(e).length > 0).length,
+              characters: characters,
+              words: words
+            };
+          }
+          return target;
+        }
+
+        let result = {};
+
+        result.rendered = processHeadings(document);
+
+        let rawHtmlDocument = getRawHtmlDocument();
+
+        if (rawHtmlDocument) {
+          result.raw = processHeadings(rawHtmlDocument);
+        }
+
+
+    
+        return result; 
+      }
+      catch(e) {
+        return logError("headings", e);
+      }
+    })(),
+
+    // Structured Data use
+    // Used by: SEO
+    // Backup: old data in almanac.js - not very good
+    'structured_data': (() => {  
+      try { 
+        var link = document.createElement('a');
+
+        function nestedJsonldLookup(target, jsonldIds, items, depth, context) {
+          if (items instanceof Array) {
+            // loop array and process any objects in it 
+            for (var i = 0, len = items.length; i < len; i++) {
+              var item = items[i];
+              if (item instanceof Object) {
+                nestedJsonldLookup(target, jsonldIds, item, depth+1, context);
+              }
+            }
+          }
+          else if (items instanceof Object) {
+            // process object
+            target.items_by_format.jsonld++;
+
+            if (items['sameAs']) { // should really check for context based entry as well.
+              // could be a string or an array
+
+              let sameAs = items['sameAs'];
+
+              if (typeof sameAs === 'string') {
+                if (!target.same_as_values.includes(sameAs)) {
+                  target.same_as_values.push(sameAs);
+                }
+              } else if (sameAs instanceof Array) {
+                sameAs.forEach((s) => {
+                  if (!target.same_as_values.includes(s)) {
+                    target.same_as_values.push(s);
+                  }
+                })
+              }             
+            }
+            
+            if (items['@id']) {
+              if (typeof items['@id'] === 'string') {
+                
+                link.href = items['@id'];
```

Comment:
Have you considered using `new URL()`?

Reply:
I copied that trick from last years code. With the URL class I'd need to pass in the base URL and test that it can work with all the weird things people enter. Too risky at this stage.

I'll put a comment in the code.